### DISCUSSION TITLE
range.hxx build failure

### DISCUSSION THREAD

**Discussion by flb-dpc** [jtv/libpqxx#849] [Q&A] [answered]
I am working in a RedHat-8 OS environment.
I discovered that configure would fail until I set the following environment variable:  CXXFLAGS="-std=c++17 -g -O2"
I configured by build using the following build options:  --disable-static --enable-shared
The configure executable is at ../libpqxx_test as one might surmise from my output below.
Configure now runs fine, but "make" produces the following failure information (absolute paths hidden by . . .):
make[1]: Leaving directory '/. . ./test'
Making all in tools
make[1]: Entering directory '/. . ./tools'
depbase=`echo rmlo.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
g++ -DHAVE_CONFIG_H   -I../include -I../../libpqxx_test/include -I/usr/include -I/usr/include  -std=c++17 -g -O2 -fvisibility=hidden -fvisibility-inlines-hidden -MT rmlo.o -MD -MP -MF $depbase.Tpo -c -o rmlo.o ../../libpqxx_test/tools/rmlo.cxx &&\
mv -f $depbase.Tpo $depbase.Po
In file included from ../../libpqxx_test/include/pqxx/pqxx:17,
                 from ../../libpqxx_test/tools/rmlo.cxx:4:
../../libpqxx_test/include/pqxx/range.hxx:149:24: error: invalid use of ‘this’ at top level
     noexcept(noexcept(*this->value() == *rhs.value()))
                        ^~~~
../../libpqxx_test/include/pqxx/range.hxx:160:24: error: invalid use of ‘this’ at top level
     noexcept(noexcept(*this == rhs))

This may be a problem with c++17, or something else.  I would be happy if there was a way to not bother making the tests and tools since I expect that I only need the libraries.  Clearly the compiler does not like how the template is being instantiated.

I am looking for ideas from the community who understands all this better than I do.

How can I configure "make" to succeed at this point?

**Comment by tt4g** [comment, marked as answer]
Duplicate #737, #750

This is GCC bug,

- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52869
- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100752

Please upgrade GCC version or use another compiler.

**Reply by flb-dpc** [reply]
Thank you very much for the prompt and valid answer.

**Reply by jtv** [reply]
By the way what version of gcc is this?  I may need to update the requirements file.

**Reply by flb-dpc** [reply]
$ gcc --version
gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-22)
Copyright (C) 2018 Free Software Foundation, Inc.

**Comment by flb-dpc** [comment]
My thanks to tt4g for the answer.
Unfortunately, the compiler version is not under my control at the moment.  I am going to need to wait for an OS upgrade to get there.
What I would have liked is a way to tell make to not build/install the tests and tools.  It is not obvious what each target in the makefile does or how I might control this using the configure step.  
What I decided to do for now is to edit the top-level Makefile from my "configure and build" script.
For those who care the important lines of bash script (with error checking not shown) that I used are:
>
    SUBDIRS_VAL="include src config"
    mv Makefile Makefile.orig
    sed -e "s/^SUBDIRS.*/SUBDIRS = ${SUBDIRS_VAL}/" Makefile.orig >Makefile
> 
Yes, I dropped "doc" on purpose also.
Thanks again to tt4g for the answer.
I am going to leave this open for a few more days in case others wish to comment.

**Reply by jtv** [reply]
The `configure` build by default does not build the tests.  And I think it builds the tools last of all, so you could probably get away with just continuing as if there were no errors.

(Also by default the latest release does not build the documentation, that's quite new.)

**Reply by jtv** [reply]
By the way I think it's not the template instantiation that the compiler can't handle, but the `noexcept` expression.  (Not the keyword per se, but the way you can "parameterise" it with an expression.)

IIRC that was a relatively recent addition to the language in C++17, and so older compilers with incomplete C++17 support may just not fully know to do that.

If so, what you can do is edit the header and remove the whole  `noexcept` specification from the function.

**Reply by flb-dpc** [reply]
When you say "latest release" I am curious where to find that?
I started my search at 
http://pqxx.org/development/libpqxx/wiki/DownloadPage
where our original version, for RedHat-6, was obtained from.  Since that no longer exists, I rolled back to
http://pqxx.org/development/libpqxx/
That page contains a lot of excellent information for developing, but I didn't need to dig into such documentation.  Looking there I found my way to GitHub where I cloned repo.
Is there actually a place that I can go to "download the latest release"?
I assume that would be source that I need to build.

**Reply by tt4g** [reply]
If you are cloning the latest source code by git, then it is the latest source code.
You can switch latest release source code by `git checkout origin 7.9.1`.

Alternatively, you can download the zip from GitHub release page: https://github.com/jtv/libpqxx/releases/tag/7.9.1

**Comment by jtv** [comment]
I'll review the documentation on libpqxx.org.

**Comment by jtv** [comment]
The documentation on libpqxx.org said to find "releases" (and other things) on github.com, and linked to the project's general github page.

For when that is not clear enough, I've added a link for just downloads.  As @tt4g said, to download a release you otherwise need to find and click the "Releases" link on the github page.

Hope that's clear enough @flb-dpc.

**Reply by flb-dpc** [reply]
It all seems clear enough in hindsight.
I believe that this stemmed from my inexperience with github in that I didn't notice, or think to look for, the "releases" link.

Working with a released zip/tar is easier because I do not need to remove all the .git stuff that has no meaning to our repository.  We work a lot here with offline systems that get "refreshed" every so-many years.  The system that I am currently "refreshing" started on RedHat-6 using version 4.0.1; that version was reused when we upgraded to RedHat-7.
Now I am working on getting things to RedHad-9, working with RedHat-8 systems for the moment.  Not sure how well 7.9.1 will work when the next refresh happens, but I will be retired by then.
Thanks for all your help.

P.S. I did try to close this discussion last week so feel free to do that for me when it makes sense to you.

**Reply by jtv** [reply]
I'm not so sure it's possible to close a Github Discussion!

**Comment by jtv** [comment]
(I've also bumped the minimum required compiler versions in `requirements.json`.)
